libraries/AC_AttitudeControl/AC_PosControl.cpp

```diff
@@ -1125,13 +1125,19 @@ void AC_PosControl::accel_to_lean_angles(float accel_x_cmss, float accel_y_cmss,
     roll_target = atanf(accel_right * cos_pitch_target / (GRAVITY_MSS * 100.0f)) * (18000.0f / M_PI);
 }
 
-// get_lean_angles_to_accel - convert roll, pitch lean angles to lat/lon frame accelerations in cm/s/s
+// get_lean_angles_to_accel - convert roll, pitch lean target angles to lat/lon frame accelerations in cm/s/s
 void AC_PosControl::lean_angles_to_accel(float& accel_x_cmss, float& accel_y_cmss) const
 {
     // rotate our roll, pitch angles into lat/lon frame
-    // todo: this should probably be based on the desired attitude not the current attitude
-    accel_x_cmss = (GRAVITY_MSS * 100) * (-_ahrs.cos_yaw() * _ahrs.sin_pitch() * _ahrs.cos_roll() - _ahrs.sin_yaw() * _ahrs.sin_roll()) / MAX(_ahrs.cos_roll() * _ahrs.cos_pitch(), 0.5f);
-    accel_y_cmss = (GRAVITY_MSS * 100) * (-_ahrs.sin_yaw() * _ahrs.sin_pitch() * _ahrs.cos_roll() + _ahrs.cos_yaw() * _ahrs.sin_roll()) / MAX(_ahrs.cos_roll() * _ahrs.cos_pitch(), 0.5f);
+    const float sin_roll = sinf(_attitude_control._attitude_target_euler_angle.x);
+    const float cos_roll = cosf(_attitude_control._attitude_target_euler_angle.x);
+    const float sin_pitch = sinf(_attitude_control._attitude_target_euler_angle.y);
+    const float cos_pitch = cosf(_attitude_control._attitude_target_euler_angle.y);
+    const float sin_yaw = sinf(_attitude_control._attitude_target_euler_angle.z);
+    const float cos_yaw = cosf(_attitude_control._attitude_target_euler_angle.z);
+
+    accel_x_cmss = (GRAVITY_MSS * 100) * (-cos_yaw * sin_pitch * cos_roll - sin_yaw * sin_roll) / MAX(cos_roll * cos_pitch, 0.5f);
```

Comment:
We currently base the acceleration vector on the real heading, not the current heading. This prevents the aircraft from toilet bowling if we have a yaw error. This happens often for multirotors and especially when we lose a motor.

I have been looking at a navigation input function that defines the thrust vector as the primary command leaving yaw as unconditional lower priority axis. However the change didn't fit well with our current code structure so I felt it would be better left for now. So we are stuck with an annoying mix of euler angle order side effects and a mix of euler and measured angles to get this right.

Reply:
I think we can pretty easily avoid the "friend class" addition by reusing get_att_target_euler_cd() in this "lean_angles_to_accel" method.  If the extra conversion to and from degrees is a problem then let's instead add another new method called "get_att_target_euler_rad()" that simply returns "_attitude_target_euler_angle".

Reply:
OK, I will remove the friend class and add the new method.